core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java

```diff
@@ -0,0 +1,812 @@
+        for (int n = 0, sz = tableNames.size(); n < sz; n++) {
+            TableUpdateDetails stats = tableUpdateDetailsByTableName.get(tableNames.get(n));
+            loadByThread[stats.threadId] += stats.nUpdates;
+        }
+    }
+
+    int[] getLoadByThread() {
+        return loadByThread;
+    }
+
+    int getnRebalances() {
+        return nRebalances;
+    }
+
+    int getnLoadCheckCycles() {
+        return nLoadCheckCycles;
+    }
+
+    @Override
+    public void close() {
+        // Both the writer and the net worker pools must have been closed so that their respective cleaners have run
+        if (null != pubSeq) {
+            pubSeq = null;
+            tableUpdateDetailsByTableName.clear();
+            for (int n = 0; n < queue.getCapacity(); n++) {
+                queue.get(n).close();
+            }
+        }
+    }
+
+    static class LineTcpMeasurementEvent implements Closeable {
+        private TruncatedLineProtoLexer lexer;
+        private final CharSequenceCache cache;
+        private final MicrosecondClock clock;
+        private final LineProtoTimestampAdapter timestampAdapter;
+        private long measurementNameAddress;
+        private final LongList addresses = new LongList();
+        private int firstFieldIndex;
+        private long timestampAddress;
+        private int errorPosition;
+        private int errorCode;
+        private int threadId;
+        private long timestamp;
+
+        private int rebalanceFromThreadId;
+        private int rebalanceToThreadId;
+        private String rebalanceTableName;
+        private boolean rebalanceReleasedByFromThread;
+
+        private LineTcpMeasurementEvent(int maxMeasurementSize, MicrosecondClock clock, LineProtoTimestampAdapter timestampAdapter) {
+            lexer = new TruncatedLineProtoLexer(maxMeasurementSize);
+            cache = lexer.getCharSequenceCache();
+            this.clock = clock;
+            this.timestampAdapter = timestampAdapter;
+            lexer.withParser(new LineProtoParser() {
+                @Override
+                public void onLineEnd(CharSequenceCache cache) {
+                }
+
+                @Override
+                public void onEvent(CachedCharSequence token, int type, CharSequenceCache cache) {
+                    assert cache == LineTcpMeasurementEvent.this.cache;
+                    switch (type) {
+                        case EVT_MEASUREMENT:
+                            assert measurementNameAddress == 0;
+                            measurementNameAddress = token.getCacheAddress();
+                            break;
+
+                        case EVT_TAG_NAME:
+                        case EVT_TAG_VALUE:
+                            assert firstFieldIndex == -1;
+                            addresses.add(token.getCacheAddress());
+                            break;
+                        case EVT_FIELD_NAME:
+                            if (firstFieldIndex == -1) {
```

Comment:
this can potentially be done branch-free, would be interesting to know if JIT does it for us